src/python/pants/backend/jvm/tasks/reports/junit_html_report.py

```diff
@@ -145,11 +145,20 @@ def as_dict(self):
     return d
 
 
-class ReportTestCase(datatype(['name', 'time', 'failure', 'error', 'skipped'])):
+class ReportTestCase(datatype([
+    'name',
+    # TODO: a `Convert` wrapper which wraps the input in float() if it's not already a float.
+    'time',
+    # TODO: an `optional(type_constraint=None)` type constraint factory which matches values
+    # matching the given constraint, as well as None.
+    ('failure', None, None),
```

Comment:
The degree of mysteriousness here is very high. The natural question from the reader is why `('failure', None, None),` instead of `'failure',`; ie: what does `'name',` above do that this does not or vice-versa? Maybe the TODO changes 'failure' in such away that `None` is no longer tossed about?

Reply:
One alternative is to use the `DatatypeFieldDecl()` constructor directly, with named arguments. One thing I did notice is that in the python 3 `typing` module, the `NamedTuple` class accepts precisely the syntax we *currently* (on `master`) have for typed fields:
```python
from typing import NamedTuple
ClassName = NamedTuple('ClassName', [('x', int)])
# versus
from pants.util.objects import datatype
class ClassName(datatype([('x', int)])): pass
# compare to:
from pants.util.objects import DatatypeFieldDecl as F
class ClassName(datatype([
    F(field_name='failure', default_value=None, has_default_value=True),
])): pass
```

So while we have actually so far kept compatible field declaration syntax with `NamedTuple`, this PR allowing mere positional tuples with this mysterious extended syntax as you've pointed out here goes past that. I think using the `DatatypeFieldDecl` constructor directly with named arguments, as above, might help this situation, and ensure we make it clear when we're writing something incompatible.